Lib/glyphsLib/casting.py

```diff
@@ -53,312 +54,469 @@
 
-def pointlist(strlist):
-    return map(point, strlist)
 
+class RWNode(RWGlyphs):
+    """Read/write a node on an outline."""
 
-def nodelist(strlist):
-    return map(node, strlist)
+    _regex = re.compile(
+        '([-.e\d]+) ([-.e\d]+) (LINE|CURVE|OFFCURVE|n/a)(?: (SMOOTH))?')
 
+    def read(self, src):
+        """Cast a node from a string with format X Y TYPE [SMOOTH]."""
+        g = self._regex.match(src).groups()
+        return [num.read(g[0]), num.read(g[1]), g[2].lower(), bool(g[3])]
 
-def glyphs_datetime(string):
-    """Parse a datetime object from a string."""
+    def write(self, val):
+        assert isinstance(val, list) and len(val) == 4
+        v2 = val[2]
+        # glyphs has this lower case
+        if v2 != 'n/a':
+            v2 = v2.upper()
+        return '%s %s %s%s' % (val[0], val[1], v2, ' SMOOTH' if val[3] else '')
 
-    # parse timezone ourselves, since %z is not always supported
-    # see: http://bugs.python.org/issue6641
-    string, tz = string.rsplit(' ', 1)
-    datetime_obj = datetime.datetime.strptime(string, '%Y-%m-%d %H:%M:%S')
-    offset = datetime.timedelta(hours=int(tz[:3]), minutes=int(tz[0] + tz[3:]))
-    return datetime_obj + offset
 
+class RWIntList(RWGlyphs):
+    """Read/write a list of ints."""
 
-def kerning(kerning_data):
-    """Cast the values in kerning data to ints."""
+    def read(self, src):
+        assert isinstance(src, list)
+        return map(int, src)
 
-    new_data = {}
-    for master_id, master_map in kerning_data.items():
-        new_data[master_id] = {}
-        for left_glyph, glyph_map in master_map.items():
-            new_data[master_id][left_glyph] = {}
-            for right_glyph, value in glyph_map.items():
-                new_data[master_id][left_glyph][right_glyph] = num(value)
-    return new_data
+    def write(self, val):
+        assert isinstance(val, list)
+        return map(str, val)
 
 
-def descender_val(string):
-    """Ensure that descender values are always negative."""
+class RWPointList(RWGlyphs):
+    """Read/write a list of points."""
 
-    num = int(string)
-    assert num < 0
-    return num
+    def read(self, src):
+        assert isinstance(src, list)
+        return map(point.read, src)
 
+    def write(self, val):
+        assert isinstance(val, list)
+        return map(point.write, val)
 
-def version_minor(string):
-    """Ensure that the minor version number is between 0 and 999."""
 
-    num = int(string)
-    assert num >= 0 and num <= 999
-    return num
+class RWNodeList(RWGlyphs):
+    """Read/write a list of nodes."""
 
+    def read(self, src):
+        assert isinstance(src, list)
+        return map(node.read, src)
 
-def normalized(string):
-    """Replace escaped characters with their intended characters.
-    Unescapes curved quotes to straight quotes, so that we can definitely
-    include this casted data in feature syntax.
-    """
+    def write(self, val):
+        return map(node.write, val)
```

Comment:
The check is now in the new utility function that modifies the list in place, which will fix this.